File: archived/dotnet_bot/TinderBot.cs
using System.Drawing;
using System.Runtime.InteropServices;
using dotnet_bot.Abstract;
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Firefox;
using OpenQA.Selenium.Interactions;
using OpenQA.Selenium.Support.UI;
using WebDriverManager;
using WebDriverManager.DriverConfigs.Impl;

namespace dotnet_bot.Bot;

public class TinderBot : BotAbstract
{
    public IWebDriver driver { get; set; }
    public string username { get; set; }
    public string password { get; set; }
    public string name { get; set; }
    public string url { get; set; }
    public string screenshotsPath { get; set; }
    public const int MAX_NUM_FILES = 20;
    private Random random = new();
    public string photoFilePath { get; set; }

    public TinderBot(
        string username,
        string password,
        string uniqueName,
        string configsPath,
        string url = "https://tinder.com"
    )
    {
        // ProxySharp.Proxy.RenewQueue();
        //var proxyServer = ProxySharp.Proxy.GetSingleProxy();
        //ProxySharp.Proxy.PopProxy();
        this.username = username;
        this.password = password;
        this.name = uniqueName;
        this.url = url;
        photoFilePath = Path.Combine(configsPath, uniqueName, $"{uniqueName}_pics");
        screenshotsPath = Path.Combine(configsPath, uniqueName, "screenshots");
        // creates the path if it doesn't exist already
        Directory.CreateDirectory(screenshotsPath);
        driver = InitFirefoxDriver();
        driver.Navigate().GoToUrl(url);
    }

    private IWebDriver InitFirefoxDriver()
    {
        FirefoxOptions firefoxOptions = new FirefoxOptions
        {
            BrowserExecutableLocation = "/usr/bin/firefox"
        };
        // firefoxOptions.SetPreference("webdriver.gecko.driver", geckoDriverPath);

        firefoxOptions.SetPreference("geo.prompt.testing", true);
        firefoxOptions.SetPreference("geo.prompt.testing.allow", true);
        firefoxOptions.SetPreference(
            "geo.provider.network.url",
            "data:application/json,{\"location\": {\"lat\": 28.538336, \"lng\": -81.379234}, \"accuracy\": 100.0}"
        );

        // FirefoxDriverService service = FirefoxDriverService.CreateDefaultService();

        // Set the path of the GeckoDriver executable
        // service.FirefoxBinaryPath = geckoDriverPath;
        // new DriverManager().SetUpDriver(new FirefoxConfig());
        // IWebDriver driver = new FirefoxDriver(firefoxOptions);
        IWebDriver driver = new FirefoxDriver(firefoxOptions);
        driver.Manage().Window.Size = new Size(1920, 1080);
        return driver;
    }

    private void print(params object[] vals) => Console.WriteLine(string.Join(",", vals));

    public override IWebElement? WaitAndGet(string xpath, float seconds = 2.5f)
    {
        try
        {
            WebDriverWait wait = new WebDriverWait(driver, TimeSpan.FromSeconds(seconds));
            var element = wait.Until(condition =>
            {
                try
                {
                    var elementToBeDisplayed = driver.FindElement(By.XPath(xpath));
                    return elementToBeDisplayed.Displayed;
                }
                catch (StaleElementReferenceException)
                {
                    return false;
                }
                catch (NoSuchElementException)
                {
                    return false;
                }
                catch (Exception)
                {
                    return false;
                }
            });

            return element ? driver.FindElement(By.XPath(xpath)) : null;
        }
        catch (Exception e)
        {
            print(e.Message);
            return null;
        }
    }

    public override IWebElement[]? WaitAndGetAll(string xpath, float seconds = 2.5f)
    {
        try
        {
            WebDriverWait wait = new WebDriverWait(driver, TimeSpan.FromSeconds(seconds));
            var element = wait.Until(condition =>
            {
                try
                {
                    var elementToBeDisplayed = driver.FindElements(By.XPath(xpath));
                    return elementToBeDisplayed.All(e => e.Displayed);
                }
                catch (StaleElementReferenceException)
                {
                    return false;
                }
                catch (NoSuchElementException)
                {
                    return false;
                }
                catch (Exception)
                {
                    return false;
                }
            });

            return element ? driver.FindElements(By.XPath(xpath)).ToArray() : null;
        }
        catch (Exception e)
        {
            print(e.Message);
            return null;
        }
    }

    public override void HandlePopups(int milliseconds = 0)
    {
        Thread.Sleep(milliseconds);

        ClickButtonWithText("Allow");

        ClickButtonWithText("Not interested");

        ClickButtonWithText("miss out");

        // handle cookie declining
        try
        {
            var declineCookieButton = WaitAndGet("//div[contains(text(), 'I decline')]");
            declineCookieButton = GetButtonParent(declineCookieButton);
            declineCookieButton?.Click();
        }
        catch (System.Exception ex)
        {
            System.Console.WriteLine(ex.Message);
        }

        ClickButtonWithText("not interested", "no thanks");

        // maybe later popup
        ClickButtonWithText("maybe later");

        // last fallback
        PressEscape();
    }

    private void PressEscape()
    {
        try
        {
            // Create an instance of Actions class
            Actions actions = new Actions(driver);

            // Send the ESCAPE key
            actions.SendKeys(Keys.Escape).Perform();
        }
        catch (System.Exception) { }
    }

    private List<string> GetMatches()
    {
        var matchProfiles = WaitAndGetAll("//a[contains(@class, 'matchListItem')]");
        var messageLinks = new List<string>();
        foreach (var profile in matchProfiles ?? [])
        {
            if (
                profile.GetAttribute("href") == "https://tinder.com/app/my-likes"
                || profile.GetAttribute("href") == "https://tinder.com/app/likes-you"
            )
            {
                continue;
            }

            messageLinks.Add(profile.GetAttribute("href"));
        }

        return messageLinks;
    }

    public override void SendMessageToMatches()
    {
        var links = GetMatches();
        foreach (var link in links)
        {
            SendMessage(link);
            Thread.Sleep(random.Next(100, 500));
        }
    }

    private void SendMessage(string link)
    {
        driver.Navigate().GoToUrl(link);
        Thread.Sleep(random.Next(4000, 5000));
        var textArea = driver.FindElement(By.TagName("textarea"));
        textArea.Clear();
        var name = driver.FindElement(By.TagName("h1"))?.Text;
        textArea.SendKeys($"nice to meet you {name ?? ""}. Hru? Got snap and/or insta? ");
        //textArea.SendKeys($"Happy new year!");

        var span = WaitAndGet("//span[contains(text(), 'Send')]");
        span = GetButtonParent(span);
        span?.Click();
    }

    public override void Screenshot()
    {
        try
        {
            string screenshotFileName = Path.Combine(
                screenshotsPath,
                $"{DateTime.Now:yyyy_MM_dd_HH_mm_ss}.png"
            );
            ((ITakesScreenshot)driver)
                .GetScreenshot()
                .SaveAsFile(screenshotFileName, ScreenshotImageFormat.Png);

            if (Directory.GetFiles(screenshotsPath).Length > MAX_NUM_FILES)
            {
                foreach (var file in Directory.GetFiles(screenshotsPath))
                {
                    File.Delete(Path.Combine(screenshotsPath, file));
                }
            }
        }
        catch
        {
            // Handle any exception
        }
    }

    public override bool TryLogin()
    {
        // count the number of attempts to login
        int tries = 0;

        bool couldLogin = GetButtonParent(WaitAndGet("//div[text()='Create account']")) is null;

        do
        {
            tries++;

            Screenshot();
            var buttons = driver.FindElements(By.TagName("button"));
            foreach (var button in buttons)
            {
                if (button is not null && button.Text.ToLower().Contains("i decline"))
                {
                    button.Click();
                    break;
                }
            }

            try
            {
                var googleAccountsPopupClose = driver.FindElement(By.Id("close"));
                googleAccountsPopupClose?.Click();
            }
            catch (System.Exception)
            {
            }

            try
            {

                // press login button
                var loginButton = WaitAndGet(
                    "//a[contains(@href,'https://tinder.onelink.me/9K8a/3d4abb81')]",
                    5
                );
                loginButton?.Click();

                var loginWithFacebook = WaitAndGet(
                    "//button[contains(@aria-label,'Log in with Facebook')]",
                    2.5f
                );
                loginWithFacebook?.Click();

                bool canContinue = false;
                var baseWindow = driver.WindowHandles[0];
                while (!canContinue)
                {
                    try
                    {
                        driver.SwitchTo().Window(driver.WindowHandles[1]);
                        canContinue = true;
                    }
                    catch (Exception)
                    {
                        canContinue = false;
                    }
                }
                // sign in
                var usernameField = WaitAndGet("//*[@id='email']", 2.5f);
                var passwordField = WaitAndGet("//*[@id='pass']", 2.5f);

                usernameField?.SendKeys(username);
                passwordField?.SendKeys(password);

                Screenshot();

                // log in
                var facebookLoginButton = WaitAndGet("//input[starts-with(@id, 'u_0_0_')]", 2.5f);
                facebookLoginButton?.Click();

                var continueAsButton = WaitAndGet("//span[contains(translate(text(), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), 'continue as')]");
                continueAsButton?.Click();

                // switch back to main window handle
                driver.SwitchTo().Window(baseWindow);

                Thread.Sleep(random.Next(5000, 10000));

                Screenshot();
            }
            catch (System.Exception)
            {

            }

            Thread.Sleep(random.Next(5000, 10000));

            couldLogin = GetButtonParent(WaitAndGet("//div[text()='Create account']")) is null;
            // System.Console.WriteLine("couldlogin: " + couldLogin);
            // System.Console.WriteLine(tries);
            if (couldLogin)
            {
                break;
            }
        }
        while (tries < 3);

        if (!couldLogin)
        {
            // System.Console.WriteLine("unable to login for " + name);
            return false;
        }

        // System.Console.WriteLine("handling pop ups rn");

        HandlePopups(random.Next(2000, 4000));

        return true;
    }

    public override bool Like()
    {
        Thread.Sleep(random.Next(500, 1500));
        try
        {
            Screenshot();
            var likeSpan = WaitAndGet("//span[text()='Like']");
            likeSpan = GetButtonParent(likeSpan);
            likeSpan?.Click();

            return true;
        }
        catch (Exception ex)
        {
            return false;
        }
    }

    public override bool NoMoreLikes()
    {
        var outOfLikes = WaitAndGet("//h3[text()='Select a plan']");

        outOfLikes ??= WaitAndGet("//h2[text()='Unlimited Likes. Send as many likes as you want.']");

        return outOfLikes is not null;
    }

    public override bool DisLike()
    {
        try
        {
            var dislikeSpan = WaitAndGet("//span[text()='Nope']");
            dislikeSpan = GetButtonParent(dislikeSpan);
            dislikeSpan?.Click();
            return true;
        }
        catch (Exception ex) {
            return false;
        }
    }

    public override bool RanOutOfMatches()
    {
        try
        {
            var outOfMatchesNotice = driver.FindElement(
                By.XPath(
                    $"//div[contains(text(),'We are unable to find any potential matches right now. Try changing your preferences to see who is nearby')]"
                )
            );
            return true;
        }
        catch (Exception ex)
        {
            return false;
        }
    }

    public override void AutoSwipe()
    {
        while (true)
        {
            //print("hello", "from", "autoswipe");
            if (!Like())
            {
                // check for no more likes pop up
                if (NoMoreLikes())
                {
                    // print("finished!");
                    break;
                }

                // check whether we ran out of matches
                if (RanOutOfMatches())
                {
                    // refresh the page
                    driver.Navigate().Refresh();
                }

                // since we still have likes left, this could have been a random popup
                PressEscape();

                HandlePopups();
            }
        }
    }

    public override IWebElement? GetButtonParent(IWebElement? e)
    {
        return base.GetButtonParent(e);
    }

    private void ClickButtonWithText(params string[] texts)
    {
        Screenshot();
        try
        {
            var buttons = driver.FindElements(By.TagName("button"));
            foreach (var button in buttons)
            {
                if (button is not null && texts.Any(text => button.Text.ToLower().Contains(text.ToLower())))
                {
                    button?.Click();
                    break;
                }
            }
        }
        catch (ElementNotInteractableException e) { }
        catch (StaleElementReferenceException e) { }
        catch (Exception ex)
        {
            System.Console.WriteLine(ex.Message);
        }
    }

    public override void Dispose()
    {
        driver.Quit();
    }

}

File: archived/dating_bot_manager/tinder_bot.py
import os
from pathlib import Path
import random as random
import time

import undetected_chromedriver as uc
from selenium.common.exceptions import (ElementNotInteractableException,
                                        NoSuchElementException,
                                        StaleElementReferenceException)
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import WebDriverWait


from bot_manager_abstract import DatingBot


class TinderBot(DatingBot):
    """Implementation of DatingBot for Tinder"""
    MAX_NUM_FILES = 20

    def __init__(self, username, password, unique_name, configs_path):
        """Initialize the WebDriver"""
        self.screenshots_path = os.path.join(configs_path, "screenshots")
        self.photo_file_path = os.path.join(configs_path, f'{unique_name}_pics')
        os.makedirs(self.screenshots_path, exist_ok=True)
        os.makedirs(self.photo_file_path, exist_ok=True)
        self.username = username
        self.password = password
        self.random = random.Random()
        options = uc.ChromeOptions()
        options.binary_location = '/snap/bin/chromium'
        options.add_argument('--private')
        options.add_argument('--disable-dev-shm-usage')
        options.add_argument('--no-sandbox')
        options.add_argument('--disable-gpu')
        options.add_argument("--incognito")
        options.add_argument("--disable-notifications")  # Disable notifications
        options.add_argument("--disable-popup-blocking")  # Disable pop-ups
        self.driver = uc.Chrome(headless=False, use_subprocess=False,)
        # self.driver.set_window_position(0, -1000)
        self.driver.implicitly_wait(5)
        self.driver.get("https://tinder.com")
        # print('hi from tinder bot')

    def wait_and_get_all(self, xpath, seconds=5):
        try:
            wait = WebDriverWait(self.driver, seconds)
            elements = wait.until(lambda driver: self._find_all_elements(xpath))
            return elements if elements else None
        except Exception as e:
            print(e)
            return None

    def wait_and_get(self, xpath, seconds=2.5):
        try:
            wait = WebDriverWait(self.driver, seconds)
            element = wait.until(lambda driver: self._find_and_display_element(xpath))
            return element if element else None
        except Exception as e:
            print(e)
            return None
    
    def _find_and_display_element(self, xpath):
        try:
            element = self.driver.find_element(By.XPATH, xpath)
            return element if element.is_displayed() else None
        except (StaleElementReferenceException, NoSuchElementException, Exception):
            return None

    def _find_all_elements(self, xpath):
        try:
            elements = self.driver.find_elements(By.XPATH, xpath)
            return elements if all(e.is_displayed() for e in elements) else None
        except (StaleElementReferenceException, NoSuchElementException, Exception):
            return None


    def screenshot(self, path=''):
        """Take a screenshot of the current browser window"""
        try:
            screenshot_pathname = os.path.join(path or self.screenshots_path, '{}.png'.format(time.strftime("%Y_%m_%d_%H_%M_%S")))
            print(screenshot_pathname)
            if self.driver:
                self.driver.save_screenshot(screenshot_pathname)
            if len(os.listdir(self.screenshots_path)) > TinderBot.MAX_NUM_FILES:
                for f in Path(self.screenshots_path).iterdir():
                    if f.is_file():
                        f.unlink()
        except Exception as e:
            pass
        

    def get_button_parent(self, e):
        if not e:
            return None
        parent = e.find_element(By.XPATH, '..')
        if parent.tag_name.lower() == 'button':
            return e
        # traverse up
        return self.get_button_parent(parent)
    
    def login(self):
        self.screenshot()
        buttons = self.driver.find_elements(By.TAG_NAME, "button")
        for button in buttons:
            if button and "i decline" in button.text.lower():
                button.click()
                break

        try:
            google_close_button = self.driver.find_element(By.ID, "close")
            google_close_button.click()
        except Exception:
            pass

        try:
            login_button = self.wait_and_get("//a[contains(@href,'https://tinder.onelink.me/9K8a/3d4abb81')]")
            login_button.click()

            facebook_button = self.wait_and_get("//button[contains(@aria-label,'Log in with Facebook')]")
            facebook_button.click()

            self.driver.switch_to.window(self.driver.window_handles[1])

            username_field = self.wait_and_get("//*[@id='email']")
            password_field = self.wait_and_get("//*[@id='pass']")

            username_field.send_keys(self.username)
            password_field.send_keys(self.password)
            self.screenshot()

            login_facebook_button = self.wait_and_get("//input[starts-with(@id, 'u_0_0_')]")
            login_facebook_button.click()

            continue_as_button = self.wait_and_get("//span[contains(text(), 'continue as')]")
            continue_as_button.click()
            time.sleep(15)
            self.screenshot()
        except Exception as e:
            pass
        print(self.driver.window_handles)
        self.driver.switch_to.window(self.driver.window_handles[0])
        self.handle_popups(3000)

    def try_login(self) -> bool:
        """Login to Tinder"""
        tries = 0
        could_login = not self.wait_and_get("//div[text()='Create account']")
        time.sleep(3)

        while tries < 3 and not could_login:
            tries += 1
            self.screenshot()
            buttons = self.driver.find_elements(By.TAG_NAME, "button")
            for button in buttons:
                if button and "i decline" in button.text.lower():
                    button.click()
                    break

            try:
                google_close_button = self.driver.find_element(By.ID, "close")
                google_close_button.click()
            except Exception:
                pass

            try:
                login_button = self.wait_and_get("//a[contains(@href,'https://tinder.onelink.me/9K8a/3d4abb81')]")
                login_button.click()

                facebook_button = self.wait_and_get("//button[contains(@aria-label,'Log in with Facebook')]")
                facebook_button.click()

                self.driver.switch_to.window(self.driver.window_handles[1])

                username_field = self.wait_and_get("//*[@id='email']")
                password_field = self.wait_and_get("//*[@id='pass']")

                username_field.send_keys(self.username)
                password_field.send_keys(self.password)
                self.screenshot()

                login_facebook_button = self.wait_and_get("//input[starts-with(@id, 'u_0_0_')]")
                login_facebook_button.click()

                continue_as_button = self.wait_and_get("//span[contains(text(), 'continue as')]")
                continue_as_button.click()

                self.driver.switch_to.window(self.driver.window_handles[0])

                time.sleep(self.random.uniform(10, 15))
                self.screenshot()

                could_login = not self.wait_and_get("//div[text()='Create account']")
            except Exception:
                pass

            time.sleep(self.random.uniform(5, 10))

        if not could_login:
            return False

        self.handle_popups(self.random.uniform(3, 4))
        return True

    def dispose(self) -> bool:
        """Close the WebDriver"""
        if self.driver:
            self.driver.quit()
            return True
        return False

    def handle_popups(self, milliseconds=0.):
        """Handle popups such as location access and notifications"""
        print("Handling popups...")
        time.sleep(milliseconds/1000)
        self.click_button_with_text('Allow')
        self.click_button_with_text('Not interested')
        self.click_button_with_text('miss out')
        try:
            decline_cookie_button = self.wait_and_get("//div[contains(text(), 'I decline')]")
            decline_cookie_button = self.get_button_parent(decline_cookie_button)
            decline_cookie_button.click()
        except Exception as e:
            print(e)
        try:
            self.click_button_with_text('not interested', 'no thanks')
        except Exception as e:
            print(e)
        self.click_button_with_text('maybe later')
        self.press_escape()
    
    def click_button_with_text(self,*args:str):
        self.screenshot()
        try:
            buttons = self.driver.find_elements(By.TAG_NAME, 'button')
            for button in buttons:
                if button and button.text.lower() in args:
                    button.click()
                    break
        except ElementNotInteractableException as e:
            print(e)
        except StaleElementReferenceException as e:
            print(e)
        except Exception as e:
            print(e)
    
    def send_message_to_matches(self):
        """Send messages to matches"""
        print("Sending messages to matches...")
    
    def like(self) -> bool:
        """Swipe right"""
        time.sleep(self.random.uniform(0.5, 1.5))
        try:
            self.screenshot()
            # like_span = self.wait_and_get("//span[text()='Like']")
            like_span = self.driver.find_element(By.XPATH, "//span[text()='Like']")
            like_span = self.get_button_parent(like_span)
            like_span.click()
            # ActionChains(self.driver).send_keys(Keys.ALT).send_keys(Keys.TAB).perform()
            return True
        except Exception as e:
            return False
        
    def dislike(self):
        """Swipe left"""
        try:
            dislike_span = self.wait_and_get("//span[text()='Nope']")
            dislike_span = self.get_button_parent(dislike_span)
            dislike_span.click()
        except Exception as e:
            pass
    
    def no_more_likes(self) -> bool:
        """Check if like limit has been reached"""
        out_of_likes = None
        try:
            out_of_likes = self.driver.find_element(By.XPATH, "//h3[text()='Select a plan']")
            # re-assign if null
            out_of_likes = self.driver.find_element(By.XPATH, "//h3[text()='Unlimited Likes. Send as many likes as you want.']") if not out_of_likes else out_of_likes
        except Exception as e:
            pass
        return out_of_likes is not None

    def ran_out_of_matches(self) -> bool:
        """Check if no more matches are available"""
        try:
            out_of_matches_notice = self.driver.find_element(By.XPATH, "//div[contains(text(),'We are unable to find any potential matches right now. Try changing your preferences to see who is nearby')]")
            return True
        except Exception as e:
            return False
    
    def press_escape(self):
        try:
            ActionChains(self.driver).send_keys(Keys.ESCAPE).perform()
        except Exception as e:
            pass

    def auto_swipe(self):
        """Automate swiping process"""
        time.sleep(30)
        print('running now!')
        while True:
            if not self.like():
                if self.no_more_likes():
                    break
                if self.ran_out_of_matches():
                    self.driver.refresh()
                # self.press_escape()
                # self.handle_popups()